test/integration/upstream_network_filter_integration_test.cc

```diff
@@ -0,0 +1,681 @@
+      lds_config_source->set_resource_api_version(envoy::config::core::v3::ApiVersion::V3);
+      auto* lds_api_config_source = lds_config_source->mutable_api_config_source();
+      lds_api_config_source->set_api_type(envoy::config::core::v3::ApiConfigSource::GRPC);
+      lds_api_config_source->set_transport_api_version(envoy::config::core::v3::V3);
+      envoy::config::core::v3::GrpcService* grpc_service =
+          lds_api_config_source->add_grpc_services();
+      setGrpcService(*grpc_service, "lds_cluster", getLdsFakeUpstream().localAddress());
+    });
+
+    BaseIntegrationTest::initialize();
+    test_server_->waitForCounterGe("listener_manager.lds.update_success", 1);
+    EXPECT_EQ(test_server_->server().initManager().state(), Init::Manager::State::Initialized);
+    registerTestServerPorts({port_name_});
+  }
+
+  void resetConnection(FakeHttpConnectionPtr& connection) {
+    if (connection != nullptr) {
+      AssertionResult result = connection->close();
+      RELEASE_ASSERT(result, result.message());
+      result = connection->waitForDisconnect();
+      RELEASE_ASSERT(result, result.message());
+      connection.reset();
+    }
+  }
+
+  ~UpstreamNetworkExtensionDiscoveryIntegrationTest() override {
+    resetConnection(ecds_connection_);
+    resetConnection(lds_connection_);
+    resetConnection(ecds2_connection_);
+  }
+
+  void createUpstreams() override {
+    BaseIntegrationTest::createUpstreams();
+    // Create the extension config discovery upstream (fake_upstreams_[1]).
+    addFakeUpstream(Http::CodecType::HTTP2);
+    // Create the listener config discovery upstream (fake_upstreams_[2]).
+    addFakeUpstream(Http::CodecType::HTTP2);
+    if (two_connections_) {
+      addFakeUpstream(Http::CodecType::HTTP2);
+    }
+  }
+
+  void waitForEcdsStream(FakeUpstream& upstream, FakeHttpConnectionPtr& connection,
+                         FakeStreamPtr& stream) {
+    AssertionResult result = upstream.waitForHttpConnection(*dispatcher_, connection);
+    ASSERT_TRUE(result);
+    result = connection->waitForNewStream(*dispatcher_, stream);
+    ASSERT_TRUE(result);
+    stream->startGrpcStream();
+  }
+
+  void waitXdsStream() {
+    // Wait for LDS stream.
+    auto& lds_upstream = getLdsFakeUpstream();
+    AssertionResult result = lds_upstream.waitForHttpConnection(*dispatcher_, lds_connection_);
+    RELEASE_ASSERT(result, result.message());
+    result = lds_connection_->waitForNewStream(*dispatcher_, lds_stream_);
+    RELEASE_ASSERT(result, result.message());
+    lds_stream_->startGrpcStream();
+
+    // Response with initial LDS.
+    sendLdsResponse("initial");
+
+    waitForEcdsStream(getEcdsFakeUpstream(), ecds_connection_, ecds_stream_);
+    if (two_connections_) {
+      // Wait for 2nd ECDS stream.
+      waitForEcdsStream(getEcds2FakeUpstream(), ecds2_connection_, ecds2_stream_);
+    }
+  }
+
+  void sendLdsResponse(const std::string& version) {
```

Comment:
It feels like there is a lot of boilerplate across these integration tests. Is there a chance to reduce some?

Reply:
There's probably some boilerplate to re-use, but I think that not a lot because the tests are running on different config areas i.e., network/listener/http/upstream/downstream.
I can open an issue an assign it on me to re-use extension discovery tests common code where possible after this one is merged

Reply:
Yes, please open an issue and see what can be done, I think tests can get away with somewhat more boilerplate than normal, but at some point it becomes a maintenance challenge and we are in danger of optimizing for write vs. read.